x-pack/plugins/case/server/client/comments/add.ts

```diff
@@ -10,135 +10,297 @@ import { pipe } from 'fp-ts/lib/pipeable';
   const query = pipe(
-    CommentRequestRt.decode(comment),
+    AlertCommentRequestRt.decode(comment),
     fold(throwErrors(Boom.badRequest), identity)
   );
 
-  decodeComment(comment);
+  decodeCommentRequest(comment);
+
+  // This function only supports adding generated alerts
+  if (comment.type !== CommentType.generatedAlert) {
+    throw Boom.internal('Attempting to add a non generated alert in the wrong context');
+  }
+  const createdDate = new Date().toISOString();
 
-  const myCase = await caseService.getCase({
+  const caseInfo = await caseService.getCase({
     client: savedObjectsClient,
-    caseId,
+    id: caseId,
   });
 
-  // An alert cannot be attach to a closed case.
-  if (query.type === CommentType.alert && myCase.attributes.status === CaseStatuses.closed) {
-    throw Boom.badRequest('Alert cannot be attached to a closed case');
+  if (
+    query.type === CommentType.generatedAlert &&
+    caseInfo.attributes.type !== CaseType.collection
+  ) {
+    throw Boom.badRequest('Sub case style alert comment cannot be added to an individual case');
   }
 
-  // eslint-disable-next-line @typescript-eslint/naming-convention
-  const { username, full_name, email } = await caseService.getUser({ request });
-  const createdDate = new Date().toISOString();
+  const userDetails: User = {
+    username: caseInfo.attributes.created_by?.username,
+    full_name: caseInfo.attributes.created_by?.full_name,
+    email: caseInfo.attributes.created_by?.email,
+  };
 
-  const [newComment, updatedCase] = await Promise.all([
-    caseService.postNewComment({
-      client: savedObjectsClient,
-      attributes: transformNewComment({
-        createdDate,
-        ...query,
-        username,
-        full_name,
-        email,
-      }),
-      references: [
-        {
-          type: CASE_SAVED_OBJECT,
-          name: `associated-${CASE_SAVED_OBJECT}`,
-          id: myCase.id,
-        },
-      ],
-    }),
-    caseService.patchCase({
-      client: savedObjectsClient,
-      caseId,
-      updatedAttributes: {
-        updated_at: createdDate,
-        updated_by: { username, full_name, email },
-      },
-      version: myCase.version,
-    }),
-  ]);
+  const subCase = await getSubCase({
+    caseService,
+    savedObjectsClient,
+    caseId,
+    createdAt: createdDate,
+    userActionService,
+    user: userDetails,
+  });
+
+  const commentableCase = new CommentableCase({
+    collection: caseInfo,
+    subCase,
+    soClient: savedObjectsClient,
+    service: caseService,
+  });
 
-  // If the case is synced with alerts the newly attached alert must match the status of the case.
-  if (newComment.attributes.type === CommentType.alert && myCase.attributes.settings.syncAlerts) {
+  const {
+    comment: newComment,
+    commentableCase: updatedCase,
+  } = await commentableCase.createComment({ createdDate, user: userDetails, commentReq: query });
+
+  if (
+    (newComment.attributes.type === CommentType.alert ||
```

Comment:
As you throw an error before if `comment.type !== CommentType.generatedAlert` it should be impossible the new comment to be of type anything other than `CommentType.generatedAlert`. Never hurts to double check :)

Reply:
haha good point.